ivy/utils/_importlib.py

```diff
@@ -1,11 +1,80 @@
+import os
 import sys
+from typing import List, Set
+from pathlib import Path
 from importlib.util import resolve_name, module_from_spec
 from ivy.utils.backend import ast_helpers
 
 
 import_cache = {}
 path_hooks = []
 
+ivy_path_abs = Path(sys.modules["ivy"].__file__).parents[1]
+
+# Note that any modules listed as 'to skip' should not depend on the Ivy backend state.
+# If they do, the behavior of ivy.with_backend is undefined and may not function as
+# Expected.
+MODULES_TO_SKIP = ["ivy.compiler"]
+
+
+def _get_modules(absolute_name: str) -> Set[str]:
+    """Get all Ivy modules and sub-packages in a module
+
+    Parameters
+    ----------
+    absolute_name
+        Full absolute name of the module
+
+    Returns
+    -------
+        A set of modules and sub-packages found
+    """
+    if not absolute_name.startswith("ivy."):
+        raise ValueError("Module name must be specified using full namespace.")
+
+    name_to_path = absolute_name.replace(".", os.path.sep)
+    module_path = os.path.join(ivy_path_abs, name_to_path)
+
+    modules = set()
+    modules.add(absolute_name)
+
+    for root, dirs, files in os.walk(module_path):
+        if root.endswith("__"):
+            continue
+        full_name = absolute_name + root[len(module_path) :].replace(os.path.sep, ".")
+        for mod_name in files:
+            if mod_name.startswith("__") or not mod_name.endswith(".py"):
+                continue
+            # [:-3] to to get the module name without .py
+            modules.add(f"{full_name}.{mod_name[:-3]}")
+        for dir_name in dirs:
+            if dir_name.startswith("__"):
+                continue
+            modules.add(f"{full_name}.{dir_name}")
+
+    return modules
+
+
+def _get_all_modules_to_skip(modules_to_skip: List[str]) -> Set[str]:
+    """Get all modules to skip during the compilation process for ivy.with_backend.
+
+    Parameters
+    ----------
+    modules
+        List of modules to skip
+
+    Returns
+    -------
+        Set of modules to skip
+    """
+    all_modules = set()
+    for module in modules_to_skip:
+        all_modules.update(_get_modules(module))
+    return all_modules
+
+
+_all_modules_to_skip = _get_all_modules_to_skip(MODULES_TO_SKIP)
+
 
```

Comment:
Since we decided to remove `MODULES_TO_EXCLUDE` I assume this is not needed anymore? since we can just check if the `absolute_name` of the module being imported contains any of the `MODULES_TO_SKIP` packages/modules e.g. if `ivy.compiler.utils` is being imported we would detect that it should be skipped because `ivy.compiler` is in `MODULES_TO_SKIP`, we would have to change the condition in the `_import_module` function tho